### PR TITLE
Fix path behaviour for set_ 

### LINKED ISSUE #213: Unexpected difference in path behavior for set_() when accessing attribute of object
Hello,

Thank you for a great library!

I came across a discrepancy when using the set_ function where using a string path and a list path behaves differently:

```
❯ python
Python 3.11.6 (main, Oct  2 2023, 13:45:54) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pydash
>>> pydash.__version__
'7.0.6'
>>> class A:
...     def __init__(self):
...             self.x = {}
...
>>> a = A()
>>> pydash.set_(a, 'x.a.b', 1)
<__main__.A object at 0x7f3e010a17d0>
>>> a.x
{'a': {'b': 1}}
>>> pydash.set_(a, ['x', 'y', 'z'], 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/venv/lib/python3.11/site-packages/pydash/objects.py", line 1825, in set_
    return set_with(obj, path, value)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/pydash/objects.py", line 1856, in set_with
    return update_with(obj, path, pyd.constant(value), customizer=customizer)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/pydash/objects.py", line 2453, in update_with
    base_set(target, last_key, callit(updater, value))
  File "/venv/lib/python3.11/site-packages/pydash/helpers.py", line 219, in base_set
    key = int(key)
          ^^^^^^^^
ValueError: invalid literal for int() with base 10: 'z'
>>> a.x
{'a': {'b': 1}, 'y': []}
```

The expected behavior would be for `a.x` to equal `{'a': {'b': 1}, 'y': {'z': 1 }}` in the end.

Note that this issue does not occur when accessing e.g. a dict directly:

```
>>> a = {}
>>> pydash.set_(a, 'x.a.b', 1)
{'x': {'a': {'b': 1}}}
>>> pydash.set_(a, ['x', 'y', 'z'], 1)
{'x': {'a': {'b': 1}, 'y': {'z': 1}}}
```

### PR DESCRIPTION
**Pull request by DeviousStoat**
fixes #213

The default type when setting a missing key is usually given by the `PathToken` built in `to_path_tokens`.
However this function is incomplete and only builds a list of `PathToken`s if the path is a string.

The issue here is that that [line](https://github.com/dgilland/pydash/blob/develop/src/pydash/objects.py#L2408) sets the default type to `list` if the object to update is not a `dict`.
And as stated above, when we have a path as a list `to_path_tokens` just returns the list and we don't have any information about the default factory, we just default to `list`.
So when we create the first list and try to update it we expect an index.

This PR updates the `to_path_tokens` function to be more general and build the list of `PathToken` even for a list.

The added `pragma: no cover` are there because this case should now be impossible to reach, the only case where the tokens are not `PathToken`s is when the path is only 1 non string value and we go directly to the end case as the for loop only iterates on the initial values.

### CODE REVIEW

File: src/pydash/objects.py [hunk 1]
```diff
@@ -2422,7 +2422,7 @@ def update_with(obj, path, updater, customizer=None):  # noqa: C901
         if isinstance(token, PathToken):
             key = token.key
             default_factory = pyd.get(tokens, [idx + 1, "default_factory"], default=default_type)
-        else:
+        else:  # pragma: no cover
```
Comment (dgilland) [inline src/pydash/objects.py:2425]: If the `else` case is no longer valid and `isinstance(token, PathToken)` is always true, then I think we could remove the if-statement and just have whatever is under the main `if` section.

Unless there was a reason to leave them in?

Comment (DeviousStoat) [reply src/pydash/objects.py]: Yeah makes sense. I removed them.

I left them just to be safe, but I guess if it happens it would probably be a rare edge case. The test cases cover a lot of cases and every passes so it should be safe enough.

**Comment by DeviousStoat** [comment]
Lint fails in CI on files I didn't touch and everything passes localy. Not sure what's going on. Maybe we can just try rerunning it?

**Comment by dgilland** [comment]
> Lint fails in CI on files I didn't touch and everything passes localy. Not sure what's going on. Maybe we can just try rerunning it?

Latest version of black made some changes to the formatting and required some flake8 ignores on certain rules. Rebase with `develop` to get those changes and lint should be good.

File: src/pydash/objects.py (hunk 1 above)
Comment (dgilland) [reply src/pydash/objects.py]: > would probably be a rare edge case

From looking at the code, the `tokens` are from `to_path_tokens()` which always returns a `PathToken` list, right? So should never be any other type. Unless I missed something in `to_path_tokens()` where a different type could be included?

Comment (DeviousStoat) [reply src/pydash/objects.py]: `to_path_tokens` can return something else than a `PathToken` if it is a single element like `"0"` of `"some_key"`.
But I guess we should probably make `to_path_tokens` always return a list of `PathToken` all the time. It would be more consistant and easier to reason about.
I added a commit with it. Let me know what you think

**Review by dgilland — APPROVED** [review]
Nice work!
